src/librustc/hir/mod.rs

```diff
@@ -444,71 +446,54 @@ pub enum TraitBoundModifier {
 pub type TyParamBounds = HirVec<TyParamBound>;
 
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
-pub struct TyParam {
-    pub name: Name,
+pub enum GenericParamKind {
+    /// A lifetime definition, eg `'a: 'b + 'c + 'd`.
+    Lifetime {
+        /// Either "'a", referring to a named lifetime definition,
+        /// or "" (aka keywords::Invalid), for elision placeholders.
+        ///
+        /// HIR lowering inserts these placeholders in type paths that
+        /// refer to type definitions needing lifetime parameters,
+        /// `&T` and `&mut T`, and trait objects without `... + 'a`.
+        name: LifetimeName,
+        bounds: HirVec<Lifetime>,
```

Comment:
Maybe we should have an `Outlives` newtype around `Lifetime` that can be used by `TyParamBound`'s variant that currently also holds `Lifetime`, by making it `Outlives(Outlives)`.

Reply:
Why not just `Outlives(Lifetime)`?